python/ray/serve/policy.py

```diff
@@ -67,11 +82,16 @@ def flush(self, endpoint_queue, backend_queues):
                 # Note(simon): This constructor takes 100+us, maybe cache this?
                 rstate = np.random.RandomState(seed)
 
-            chosen_backend = rstate.choice(
-                self.backend_names, replace=False,
-                p=self.backend_weights).squeeze()
+            chosen_backend, shadow_backends = self._select_backends(
+                rstate.random())
 
             assigned_backends.add(chosen_backend)
             backend_queues[chosen_backend].add(query)
+            if len(shadow_backends) > 0:
+                shadow_query = query.copy()
```

Comment:
this might not copy the asyncio future, can you double check and make sure the query.async_future is pointing to different objects? if they point to the same, bad things will happen (asyncio.future is first writer win)

Reply:
The async future isn't actually used. I will just set it to None in the copy